docs/proposals/05-object-storage.md

```diff
@@ -0,0 +1,146 @@
+A `Bucket` is the S3 enabled storage endpoint. IOPS/Bandwidth limitations are controlled via a `BucketClass` and the 
+capabilities of the underlying storage provider are expressed via a `BucketPool`. A `Bucket` can be requested 
+from a `BucketPool` as long as it can provide the performance characteristics described in the `BucketClass`. 
+The proposed API resources are similar to `Volume`, `VolumeClass` and `Volumepool` except that a volume is a 
+block device with a specific driver.
+
+
+### Bucket
+
+A `Bucket` is a *namespaced* resource to request S3-compatible object storage. 
+`Bucket`s with the `type` `Public` set (only valid type for now) are accessible from the internet.
+The desired `BucketClass` is referenced by the `bucketClassRef`. If no pool is pre-defined, 
+the `bucketPoolSelector` will be used to find a suitable `BucketPool`.  The desired pool, either pre-defined or 
+set by another controller, is stated in the `bucketPoolRef`.
+
+The information to access the requested `Bucket` is in the `access` field of the status. 
+The `endpoint` defines the address of the `Bucket` Rest-API. Access credentials are placed in a secret with is referenced 
+through the `secretRef`. The `state` indicates if the `Bucket` is `Available`, `Pending` or in an `Error` state.
+
+[//]: # (@formatter:off)
+```yaml
+apiVersion: storage.api.onmetal.de/v1alpha1
+kind: Bucket
+metadata:
+  name: bucket-1
+spec:
+  # currently only type public is defined
+  type: Public
+  bucketClassRef:
+    name: slow
+  bucketPoolSelector:
+    matchLabels:
+      key: db
+      foo: bar
+  bucketPoolRef:
+    name: fra-shared
+status:
+  access:
+    endpoint: foo.bar.example.org 
+    secretRef:
+      name: 000225194345f27a40257c5777c96a03ce219f96731f22afc45b7dfda7d077d
+  state: Available
+```
+[//]: # (@formatter:on)
+
+### BucketClass
+
+A `BucketClass` is a *non-namespaced* resource which describes the characteristics of a `Bucket`. The maximal 
+performance the `Bucket` will offer (like I/O operations or throughput) is defined in the `capabilities` field.
+
+[//]: # (@formatter:off)
+```yaml
+apiVersion: storage.api.onmetal.de/v1alpha1
+kind: BucketClass
+metadata:
+  name: slow
+capabilities:
+  iops: 10
+  tps: 20Mi
+```
+[//]: # (@formatter:on)
+
+### BucketPool
+
+A `BucketPool` is a *non-namespaced* logical unit and accommodates a collection of `Bucket`s. 
+The `provider`'s id (the implementor's id) of the `BucketPool` is stated in the `providerID` field. 
+Only `Bucket`s who tolerate all the taints, will land in the `BucketPool`. `BucketClasses` which can be fulfilled by 
+the provider of the `BucketPool`, are listed in the status field `availableBucketClasses`. The `state` in the status 
+indicates if the pool is `Available`, `Pending` or `Unavailable`.
+
+[//]: # (@formatter:off)
+```yaml
+apiVersion: storage.api.onmetal.de/v1alpha1
+kind: BucketPool
+metadata:
+  name: ceph-object-store
+spec:
+  providerID: cephlet://pool
```

Comment:
can we hide `ceph` in our APIs? We should use something else als providerID.

Reply:
The providerID is basically a reference to the implementor of the pool. In our case it's `ceph` but could be also for instance `min.io`. Do you propose to remove it completely or just use a different sample?

Reply:
Our customers shouldn't care about what storage provider we use (here: Ceph). We should expose our storage provider as something like "onmetal-object-storage" or "onmetal-block-storage", so we can replace Ceph with a different provider if we need to without customer impact.

Reply:
In our lowest cluster we will find a `BucketPool` like in the sample above. In our case the `cephlet` will create this resource and sets it own provider id. Through syncing the `BucketPool` up to the customer cluster the `providerID` will be replaced through the `id` from one layer below.

Reply:
It's the same mechanism like for the `VolumePools`. In the root cluster we have a `VolumePool`:
```
➜  ~ k get volumepools fra3 -o yaml 
apiVersion: storage.api.onmetal.de/v1alpha1
kind: VolumePool
metadata:
  name: fra3
spec:
  providerID: partition://fra3
status:
  ...
  state: Available
  ``` 

And in the lowest cluster we can identify the actual provider: 
```
apiVersion: storage.api.onmetal.de/v1alpha1
kind: VolumePool
metadata:
  name: ceph
spec:
  providerID: cephlet://pool
status:
  availableVolumeClasses:
   ...
  state: Available
  ```

Reply:
I agree here. But currently this is an additional feature in the pool. So the providerID is just 'who cares about this pool' ... i guess for the moment those names are not to be valued too much - cephlet is an operator and exposes somewhat the relation to ceph, but storagelet was also a possible name, that would conceal this better. I would not spend to much time on provider names ... will probably change anyhow.

Reply:
It probably doesn't matter for the spec (and is no blocker for this proposal) but we should still get this right.

Reply:
I assume this conversation is also resolved.